### PR TITLE
fix: find common header between the header chain and the current body chain

### PR DESCRIPTION
**Pull request by garyyu**
Fix of #3030 

The current `is_on_current_chain()` function has an implicit meaning of **current `header` chain**, don't remember when we did this change, anyway it impacts the logic here to find the header common between the header chain and the current body chain.

### CODE REVIEW

**Comment by antiochp** [comment]
> The current is_on_current_chain() function has an implicit meaning of current header chain, don't remember when we did this change, anyway it impacts the logic here to find the header common between the header chain and the current body chain.

I don't fully understand what you mean here by "implicit meaning of current header chain".
This is the _only_ way of knowing if we are on the current chain or not based on our header MMR.

We check for `is_on_current_chain` with the following logic - 

* put header MMR in correct state via `rewind_and_reapply_header_fork`
  * header MMR data file now contains sequential headers up to and including current height
  * this may have diverged from the files on disk if this is a fork (we only persist the most work chain)
* for the header in question
  * take its height and lookup the header in our MMR at this height
  * if its the same header its on the "current" chain
  * otherwise it is on a different fork

i.e. If block `abc` has height 1,000 and on the current chain the header at height 1,000 is `abc` then we know `abc` is on the current chain.
Which feels a little circular but if we find block `def` at height 1,000 then we know `abc` cannot also be at height 1,000 and _must_ be on a different fork.

File: chain/src/chain.rs [hunk 1]
```diff
@@ -791,11 +791,11 @@ impl Chain {
 		let mut oldest_height = 0;
 		let mut oldest_hash = ZERO_HASH;
 
-		let mut current = self.get_block_header(&header_head.last_block_h);
```
Comment (antiochp) [inline chain/src/chain.rs]: I don't think this change makes sense.
We are checking to see if a `header` exists in the db and we should do this based on the current `header_head` which is the tip of the current header chain.

We are in `check_txhashset_needed` which (confusingly) returns a vec of block hashes to go request (if we do not need to request a txhashset archive).
So we need to determine which blocks we do not have, based on the headers we do have. So we should be looking at the headers in both our db and our header MMR. Headers up to and including the current tip of the header chain (`header_head`).

File: chain/src/chain.rs [hunk 2]
```diff
@@ -807,34 +807,41 @@ impl Chain {
...
-			oldest_hash = header.hash();
-			if let Some(hs) = hashes {
-				hs.push(oldest_hash);
-			}
 			current = self.get_previous_header(&header);
 		}
 
+		// Go through the header chain reversely to get all those headers after oldest_height.
+		if let Some(hs) = hashes {
+			let mut current = self.get_block_header(&header_head.last_block_h);
```
Comment (antiochp) [inline chain/src/chain.rs]: I see we are redefining current in this PR here, but I don't think we need to do this. Earlier this is `body_head` and now this is redefined based on `header_head`. 

I think it should consistently be based on a header from the db starting at `header_head` and working back until we find one that also refers to a full block in our db.

File: chain/src/chain.rs [hunk 3]
```diff
@@ -791,11 +791,11 @@ impl Chain {
 		let mut oldest_height = 0;
 		let mut oldest_hash = ZERO_HASH;
 
-		let mut current = self.get_block_header(&header_head.last_block_h);
+		let mut current = self.get_block_header(&body_head.last_block_h);
```
Comment (antiochp) [inline chain/src/chain.rs]: Lets rename _this_ usage of current to make it clearer that we want to check for headers in our db based on `body_head`.

This is a 2 step process.

* start with `head` (head of the full block chain)
  * if we do not have a _full block_ in the db for this `head` then its an error
  * if we do not have the header in the db for this `head` then its an error
* traverse back through the full block chain from `head` until we find a header that "is on current chain"
  * this is our "fork point" between existing header chain and full block chain
* now we want to traverse back through our header chain from `header_head` back to this fork point
  * these are the blocks that we need to request (we have the header but not the full block)

Comment (garyyu) [reply chain/src/chain.rs]: I put some of yours comments to the source code comments, and renamed the 2nd `current`. 
pls let me whether it's better now.

**Pull request merged**